### ISSUE TITLE
Support TMS indexing in TileLayer

### ISSUE THREAD

**Issue by uppeabhishek** [visgl/deck.gl#4757]
I'm using MVT Layer in deckgl with GeoServer and when the MVT layer is trying to fetch the data from the url the inverted y axis 
i.e. -y goes undefined.

```
new deck.DeckGL({
                initialViewState: {
                    longitude: -79.3832,
                    latitude: 43.6532,
                    zoom: 14
                },
                controller: true,
                layers: [
                    new deck.MVTLayer({
                        data: "/geoserver/gwc/service/tms/1.0.0/"+layer+"@EPSG%3A"+projection_epsg_no+"@pbf/{z}/{x}/{-y}.pbf",
                    })
                ]
            })
```
For the above code MVT layer is considering -y as undefined as below.

/geoserver/gwc/service/tms/1.0.0/ibi:SWwidths_v1@EPSG%3A900913@pbf/14/4579/**undefined**.pbf.

I even tried using open layers with the same URL and it was considering -y properly.

**Comment by ibesora** [comment]
Wouldn't setting the - not as a param but as the url string work?
`data: "/geoserver/gwc/service/tms/1.0.0/"+layer+"@EPSG%3A"+projection_epsg_no+"@pbf/{z}/{x}/-{y}.pbf"`

I think deck.gl is looking for {x}, {y} and {z} to substitute values, that's why {-y} results in undefined

**Comment by kylebarron** [comment]
Basically you're trying to request TMS tiles with the TileLayer? 

I don't think this will work with the existing TileLayer without a change. Even when you put the `-` outside of the `{y}`, the issue is that TMS tiles are in a different location than the standard XYZ tiles, so deck.gl would try to render them in the wrong place.

**Comment by kylebarron** [comment]
The difference between XYZ and TMS indexing is a flipped `y` coordinate, see: https://gist.github.com/tmcw/4954720.

To support this we could have a `tms` prop on the TileLayer, and internally call `y = Math.pow(2, z) - y - 1;` when the flag is true.

**Comment by ibgreen** [comment]
That sounds reasonable. Are there perhaps additional indexing schemes we might want to support?

```
indexing`: 'tms'
```
vs
```
tms: true
```

**Comment by kylebarron** [comment]
XYZ and TMS indexing are really the only two _for the web mercator grid_. The only difference is one's origin is the bottom left and one's is the top left. Only one line of code is necessary because while indexing is slightly different, the tiles are in the same place.

However there are plenty of other _non-web mercator grids_, which would be a little more involved to implement because the positions of the tiles are different. For example, at some point I want to see if it's easy to support the global geodetic tiling system (what Cesium uses), where the tiles use unprojected, geographic positions. (two tiles at zoom 0, each a square spanning 180 degrees of latitude/longitude; and a quadtree within each). I think essentially you'd only have to subclass `OSMNode` and change `getBoundingVolume`.
https://github.com/visgl/deck.gl/blob/03a6e650e4ae69fffe59965f188ea3808c93102c/modules/geo-layers/src/tile-layer/tile-2d-traversal.js#L78-L89

**Comment by Pessimistress** [comment]
Could we just add support for `{-y}` in the URL template? Evaluates to `Math.pow(2, z) - y - 1`

**Comment by kylebarron** [comment]
The question is when is that resolved. If someone passes that in the `data` prop, but has a custom `getTileData`, then it would be incorrect to evaluate it before `getTileData`, because they'll fetch the wrong data.

**Comment by Pessimistress** [comment]
The resolved tile url is passed to `getTileData`: https://deck.gl/docs/api-reference/geo-layers/tile-layer#gettiledata

**Comment by kylebarron** [comment]
Right, and I'm saying that's wrong in this case. The point is that the _server_ is what's requiring TMS.

If the value of `data` is `https://example.com/{z}/{x}/{-y}.png`, then for TMS tile `x=1, y=1, z=3`, that would be resolved as `x=1, y=6, z=3`, which is the right XYZ location. But `getTileData` would receive the resolved _XYZ index_
```
https://example.com/3/1/6.png
```
when it instead needs to receive the _TMS index_
```
https://example.com/3/1/1.png
```

**Comment by Pessimistress** [comment]
If `data` is `https://example.com/{z}/{x}/{-y}.png` then `getTileData` will receive:

```js
{
  x: 1,
  y: 6,
  z: 3,
  url: `https://example.com/3/1/1.png`
}
```

True that the user will need to understand that `y` is not from the TMS indexing system if they want to handle this information programmatically. However, as far as I can tell, the original issue is about the automatic URL resolution.

**Comment by kylebarron** [comment]
I think that's a valid change at least for now.

**Comment by Pessimistress** [comment]
This is now available in 8.3.0-beta
